File: near/src/main/java/com/adroitandroid/near/discovery/client/BroadcastHandler.java
package com.adroitandroid.near.discovery.client;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.Enumeration;

public class BroadcastHandler extends Handler {
    static final int STOP_BROADCAST = 1235;
    static final int REPEAT_BROADCAST = 2345;
    private final String mHostName;
    private final long mBroadcastInterval;
    private DatagramSocket mSocket;

    BroadcastHandler(Looper looper, String hostName, long broadcastInterval) {
        super(looper);
        mHostName = hostName;
        mBroadcastInterval = broadcastInterval;
        try {
            mSocket = new DatagramSocket();
            mSocket.setBroadcast(true);
        } catch (SocketException ignored) { }
    }

    @Override
    public void handleMessage(@NonNull Message msg) {
        if (msg.what == REPEAT_BROADCAST) {
            broadcast();
        } else if (msg.what == STOP_BROADCAST) {
            removeMessages(REPEAT_BROADCAST);
            mSocket.close();
        }
    }

    void broadcast() {
        if (mSocket != null) {
            try {
                byte[] sendData = mHostName.getBytes();
                try {
                    DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length,
                            InetAddress.getByName("255.255.255.255"), 8888);
                    mSocket.send(sendPacket);
                } catch (Exception ignored) { }

                Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
                while (interfaces.hasMoreElements()) {
                    NetworkInterface networkInterface = interfaces.nextElement();

                    if (networkInterface.isLoopback() || !networkInterface.isUp()) {
                        continue;
                    }

                    for (InterfaceAddress interfaceAddress : networkInterface.getInterfaceAddresses()) {
                        InetAddress broadcast = interfaceAddress.getBroadcast();
                        if (broadcast == null) {
                            continue;
                        }

                        try {
                            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, broadcast, 8888);
                            mSocket.send(sendPacket);
                        } catch (Exception ignored) { }
                    }
                }
                sendEmptyMessageDelayed(REPEAT_BROADCAST, mBroadcastInterval);
            } catch (IOException ignored) { }
        }
    }
}

File: near/src/main/java/com/adroitandroid/near/discovery/client/BroadcastHandler.kt
package com.adroitandroid.near.discovery.client

import android.os.Handler
import android.os.Looper
import android.os.Message
import org.json.JSONObject
import java.io.IOException
import java.net.*

class BroadcastHandler internal constructor(looper: Looper,
                                            private val mHostJson: JSONObject,
                                            private val mBroadcastInterval: Long,
                                            private val port: Int): Handler(looper) {
    private var mSocket: DatagramSocket? = null

    override fun handleMessage(msg: Message) {
        if (msg.what == REPEAT_BROADCAST) {
            broadcast()
        } else if (msg.what == STOP_BROADCAST) {
            removeMessages(REPEAT_BROADCAST)
            mSocket!!.close()
        }
    }

    fun broadcast() {
        if (mSocket != null) {
            try {
                val sendData = mHostJson.toString().toByteArray()
                try {
                    val sendPacket = DatagramPacket(sendData, sendData.size,
                            InetAddress.getByName("255.255.255.255"), port)
                    mSocket!!.send(sendPacket)
                } catch (ignored: Exception) {
                }
                val interfaces = NetworkInterface.getNetworkInterfaces()
                while (interfaces.hasMoreElements()) {
                    val networkInterface = interfaces.nextElement()
                    if (networkInterface.isLoopback || !networkInterface.isUp) {
                        continue
                    }
                    for (interfaceAddress in networkInterface.interfaceAddresses) {
                        val broadcast = interfaceAddress.broadcast ?: continue
                        try {
                            val sendPacket = DatagramPacket(sendData, sendData.size, broadcast, port)
                            mSocket!!.send(sendPacket)
                        } catch (ignored: Exception) {
                        }
                    }
                }
                sendEmptyMessageDelayed(REPEAT_BROADCAST, mBroadcastInterval)
            } catch (ignored: IOException) {
            }
        }
    }

    companion object {
        const val STOP_BROADCAST = 1235
        const val REPEAT_BROADCAST = 2345
    }

    init {
        try {
            mSocket = DatagramSocket()
            mSocket!!.broadcast = true
        } catch (ignored: SocketException) { }
    }
}